crates/walrus-sdk/src/client/byte_range_read_client.rs

```diff
@@ -0,0 +1,479 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Client for reading byte ranges from blobs.
+
+use std::{num::NonZeroUsize, time::Duration};
+
+use serde::{Deserialize, Serialize};
+use tokio::time::Instant;
+use walrus_core::{
+    BlobId,
+    Epoch,
+    SliverIndex,
+    encoding::{EncodingFactory, Primary, SliverData},
+    metadata::{BlobMetadataApi as _, VerifiedBlobMetadataWithId},
+};
+use walrus_sui::client::ReadClient;
+
+use crate::{
+    client::WalrusNodeClient,
+    error::{ClientError, ClientErrorKind, ClientResult},
+};
+
+/// Configuration for the ByteRangeReadClient.
+#[serde_with::serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct ByteRangeReadClientConfig {
+    /// The maximum number of attempts to retrieve slivers.
+    pub max_retrieve_slivers_attempts: usize,
+    /// The timeout duration for retrieving slivers.
+    #[serde_as(as = "serde_with::DurationSeconds")]
+    #[serde(rename = "timeout_secs")]
+    pub timeout: Duration,
+}
+
+impl ByteRangeReadClientConfig {
+    /// Creates a new ByteRangeReadClientConfig.
+    pub fn new(max_retrieve_slivers_attempts: usize, timeout: Duration) -> Self {
+        Self {
+            max_retrieve_slivers_attempts,
+            timeout,
+        }
+    }
+}
+
+impl Default for ByteRangeReadClientConfig {
+    fn default() -> Self {
+        Self {
+            max_retrieve_slivers_attempts: 2,
+            timeout: Duration::from_secs(10),
+        }
+    }
+}
+
+/// A client for reading specific byte ranges from blobs.
+#[derive(Debug, Clone)]
+pub struct ByteRangeReadClient<'a, T> {
+    client: &'a WalrusNodeClient<T>,
+    config: ByteRangeReadClientConfig,
+}
+
+impl<'a, T> ByteRangeReadClient<'a, T> {
+    /// Creates a new ByteRangeReadClient.
+    pub fn new(client: &'a WalrusNodeClient<T>, config: ByteRangeReadClientConfig) -> Self {
+        Self { client, config }
+    }
+}
+
+impl<T: ReadClient> ByteRangeReadClient<'_, T> {
+    /// Reads a specific byte range from a blob.
+    pub async fn read_byte_range(
+        &self,
+        blob_id: &BlobId,
+        start_byte_position: usize,
+        byte_length: usize,
+    ) -> ClientResult<Vec<u8>> {
+        // To read the byte range of the original blob file, we find the corresponding primary
+        // slivers that covers the byte range, and retrieve the slivers from the Walrus shards,
+        // and extract the requested data from the slivers.
+
+        tracing::debug!(
+            %blob_id,
+            start_byte_position,
+            byte_length,
+            "start reading byte range"
+        );
+
+        let Some(byte_length) = NonZeroUsize::new(byte_length) else {
+            return Ok(vec![]);
+        };
+
+        // Validate blob ID
+        self.client.check_blob_id(blob_id)?;
```

Comment:
I think `check_blob_id` should be done before we do byte length check. For a byte length of 0, we will return Ok(vec![]) but if it is for a blob in blocklist, it should ideally see an error with correct error code. This also keeps the behavior consistent with full blob read.

Reply:
Good call! I overlooked what `check_blob_id` actually did. Your suggestion is important. Done.